File: ros2/tutorials/lifecycle/src/my_lifecycle_py/my_lifecycle_py/lifecycle_manager.py
# Copyright 2024 Developer
# SPDX-License-Identifier: Apache-2.0
"""Simple lifecycle manager that drives another node through its states."""

import time
import rclpy
from rclpy.node import Node
from lifecycle_msgs.srv import ChangeState, GetState
from lifecycle_msgs.msg import Transition

TRANSITIONS = {
    'configure':  Transition.TRANSITION_CONFIGURE,
    'activate':   Transition.TRANSITION_ACTIVATE,
    'deactivate': Transition.TRANSITION_DEACTIVATE,
    'cleanup':    Transition.TRANSITION_CLEANUP,
    'shutdown':   Transition.TRANSITION_UNCONFIGURED_SHUTDOWN,
}


class LifecycleManager(Node):
    """Drives a target lifecycle node through a predefined sequence."""

    def __init__(self, target_node: str = 'lifecycle_talker') -> None:
        super().__init__('lifecycle_manager')
        self._target = target_node
        # Create service clients
        self._change_state = self.create_client(
            ChangeState, f'/{target_node}/change_state')
        self._get_state = self.create_client(
            GetState, f'/{target_node}/get_state')
        self.get_logger().info(
            f'LifecycleManager ready to manage: {target_node}')

    def get_state(self) -> str | None:
        """Return the current primary state label of the managed node."""
        if not self._get_state.wait_for_service(timeout_sec=5.0):
            self.get_logger().error('get_state service not available')
            return None
        future = self._get_state.call_async(GetState.Request())
        rclpy.spin_until_future_complete(self, future, timeout_sec=5.0)
        if future.result():
            return future.result().current_state.label
        return None

    def change_state(self, transition_name: str) -> bool:
        """Trigger a named transition. Returns True on success."""
        if transition_name not in TRANSITIONS:
            self.get_logger().error(f'Unknown transition: {transition_name}')
            return False
        if not self._change_state.wait_for_service(timeout_sec=5.0):
            self.get_logger().error('change_state service not available')
            return False
        req = ChangeState.Request()
        req.transition.id = TRANSITIONS[transition_name]
        req.transition.label = transition_name
        future = self._change_state.call_async(req)
        rclpy.spin_until_future_complete(self, future, timeout_sec=10.0)
        if future.result() and future.result().success:
            state = self.get_state()
            self.get_logger().info(
                f'Transition "{transition_name}" succeeded. New state: {state}')
            return True
        else:
            self.get_logger().error(f'Transition "{transition_name}" FAILED.')
            return False

    def run_demo_sequence(self) -> None:
        """Run configure -> activate -> (wait) -> deactivate -> cleanup."""
        steps = [
            ('configure', 2.0),
            ('activate',  5.0),
            ('deactivate', 1.0),
            ('cleanup',   1.0),
        ]
        for transition, wait in steps:
            if not self.change_state(transition):
                self.get_logger().error(f'Aborting sequence at: {transition}')
                return
            time.sleep(wait)
        self.get_logger().info('Demo sequence complete.')


def main(args=None):
    rclpy.init(args=args)
    manager = LifecycleManager(target_node='lifecycle_talker')
    # Wait briefly for the talker to come up
    time.sleep(2.0)
    manager.run_demo_sequence()
    manager.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2/tutorials/lifecycle/src/my_lifecycle_cpp/src/lifecycle_manager.cpp
// Copyright 2024 Developer
// SPDX-License-Identifier: Apache-2.0
/**
 * @brief Simple lifecycle manager that orchestrates another lifecycle node.
 *
 * Calls /<target>/change_state and /<target>/get_state services.
 */
#include <chrono>
#include <memory>
#include <string>
#include <thread>
#include "rclcpp/rclcpp.hpp"
#include "lifecycle_msgs/srv/change_state.hpp"
#include "lifecycle_msgs/srv/get_state.hpp"
#include "lifecycle_msgs/msg/transition.hpp"

using namespace std::chrono_literals;
using ChangeState = lifecycle_msgs::srv::ChangeState;
using GetState    = lifecycle_msgs::srv::GetState;

class LifecycleManager : public rclcpp::Node
{
public:
  explicit LifecycleManager(const std::string & target = "lifecycle_talker")
  : Node("lifecycle_manager"), target_(target)
  {
    change_state_client_ = create_client<ChangeState>("/" + target + "/change_state");
    get_state_client_    = create_client<GetState>("/" + target + "/get_state");
    RCLCPP_INFO(get_logger(), "LifecycleManager ready to manage: %s", target_.c_str());
  }

  std::string get_state()
  {
    if (!get_state_client_->wait_for_service(5s)) {
      RCLCPP_ERROR(get_logger(), "get_state service unavailable");
      return "unknown";
    }
    auto req = std::make_shared<GetState::Request>();
    auto future = get_state_client_->async_send_request(req);
    if (rclcpp::spin_until_future_complete(
        get_node_base_interface(), future, 5s) == rclcpp::FutureReturnCode::SUCCESS)
    {
      return future.get()->current_state.label;
    }
    return "unknown";
  }

  bool change_state(uint8_t transition_id, const std::string & label)
  {
    if (!change_state_client_->wait_for_service(5s)) {
      RCLCPP_ERROR(get_logger(), "change_state service unavailable");
      return false;
    }
    auto req = std::make_shared<ChangeState::Request>();
    req->transition.id    = transition_id;
    req->transition.label = label;
    auto future = change_state_client_->async_send_request(req);
    if (rclcpp::spin_until_future_complete(
        get_node_base_interface(), future, 10s) == rclcpp::FutureReturnCode::SUCCESS)
    {
      if (future.get()->success) {
        RCLCPP_INFO(get_logger(), "Transition '%s' succeeded. New state: %s",
          label.c_str(), get_state().c_str());
        return true;
      }
    }
    RCLCPP_ERROR(get_logger(), "Transition '%s' FAILED.", label.c_str());
    return false;
  }

  // -- Convenience wrappers ---------------------------------------------------
  bool configure()  { return change_state(lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE,  "configure");  }
  bool activate()   { return change_state(lifecycle_msgs::msg::Transition::TRANSITION_ACTIVATE,   "activate");   }
  bool deactivate() { return change_state(lifecycle_msgs::msg::Transition::TRANSITION_DEACTIVATE, "deactivate"); }
  bool cleanup()    { return change_state(lifecycle_msgs::msg::Transition::TRANSITION_CLEANUP,    "cleanup");    }
  bool shutdown()   { return change_state(lifecycle_msgs::msg::Transition::TRANSITION_UNCONFIGURED_SHUTDOWN, "shutdown"); }

  void run_demo()
  {
    std::this_thread::sleep_for(2s);  // wait for talker to start
    if (!configure())  return;
    std::this_thread::sleep_for(1s);
    if (!activate())   return;
    std::this_thread::sleep_for(5s);  // let it publish for 5 s
    if (!deactivate()) return;
    std::this_thread::sleep_for(1s);
    if (!cleanup())    return;
    RCLCPP_INFO(get_logger(), "Demo sequence complete.");
  }

private:
  std::string target_;
  rclcpp::Client<ChangeState>::SharedPtr change_state_client_;
  rclcpp::Client<GetState>::SharedPtr    get_state_client_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto manager = std::make_shared<LifecycleManager>("lifecycle_talker");
  manager->run_demo();
  rclcpp::shutdown();
  return 0;
}
